app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationsPagingSource.kt

```diff
@@ -0,0 +1,184 @@
+
+        try {
+            val response = when (params) {
+                is LoadParams.Refresh -> {
+                    getInitialPage(params)
+                }
+                is LoadParams.Append -> mastodonApi.notifications(
+                    maxId = params.key,
+                    limit = params.loadSize,
+                    excludes = notificationFilter
+                )
+                is LoadParams.Prepend -> mastodonApi.notifications(
+                    minId = params.key,
+                    limit = params.loadSize,
+                    excludes = notificationFilter
+                )
+            }
+
+            if (!response.isSuccessful) {
+                return LoadResult.Error(Throwable(response.errorBody().toString()))
+            }
+
+            val links = getPageLinks(response.headers()["link"])
+            return LoadResult.Page(
+                data = response.body()!!,
+                nextKey = links.next,
+                prevKey = links.prev
+            )
+        } catch (e: Exception) {
+            return LoadResult.Error(e)
+        }
+    }
+
+    /**
+     * Fetch the initial page of notifications, using params.key as the ID of the initial
+     * notification to fetch.
+     *
+     * - If there is no key, a page of the most recent notifications is returned
+     * - If the notification exists, and is not filtered, a page of notifications is returned
+     * - If the notification does not exist, or is filtered, the page of notifications immediately
+     *   before is returned
+     * - If there is no page of notifications immediately before then the page immediately after
+     *   is returned
+     */
+    private suspend fun getInitialPage(params: LoadParams<String>): Response<List<Notification>> = coroutineScope {
+        // If the key is null this is straightforward, just return the most recent notifications.
+        val key = params.key
+            ?: return@coroutineScope mastodonApi.notifications(
+                limit = params.loadSize,
+                excludes = notificationFilter
+            )
+
+        // It's important to return *something* from this state. If an empty page is returned
+        // (even with next/prev links) Pager3 assumes there is no more data to load and stops.
+        //
+        // In addition, the Mastodon API does not let you fetch a page that contains a given key.
+        // You can fetch the page immediately before the key, or the page immediately after, but
+        // you can not fetch the page itself.
+
+        // First, try and get the notification itself, and the notifications immediately before
+        // it. This is so that a full page of results can be returned. Returning just the
+        // single notification means the displayed list can jump around a bit as more data is
+        // loaded.
+        //
+        // Make both requests, and wait for the first to complete.
```

Comment:
Make two requests parallely?

Reply:
I'm not sure what you mean with this comment.

Reply:
It ties in with my complexity observation. It is a bit suprising and I'd expect it to be difficult to "join" the results of two requests.

But, yes, resolve it.

Reply:
Will do.

Joining multiple network calls with Kotlin is pretty straightforward.

If you want to wait for calls to complete before proceeding you can use the `zip` operator, there's a good introduction to that at https://amitshekhar.me/blog/kotlin-flow-zip-operator-parallel-multiple-network-calls.

If you need the result of one or more them before you need the result of another you do exactly what's done here -- launch them with `async`,  then call `.await()` on whichever one you need the value for first.